docs/design.md

```diff
@@ -1,87 +1,133 @@
-
-It would be better if CSI driver could get reliable and fast event from AWS when a volume has become attached / detached.
-
-Or the driver could batch the calls and issue one big DescribeVolume call with every volume that&#39;s being attached/detached in it.
-
-### AWS API weirdness
-
-AWS API is quite different to all other clouds.
-
-- AWS.AttachVolume/DetachVolume can take ages (days, weeks) to complete. For example, when Kubernetes tries to detach a volume that&#39;s still mounted on a node, it will be Detaching until the volume is unmounted or Kubernetes issues force-detach. All other clouds return sane error relatively quickly (e.g. &quot;volume is in use&quot;) or force-detach the volume.
-- AWS.DetachVolume with force-detach is useless. Documentation says: Forced detachment of a stuck volume can cause damage to the file system or the data it contains or an inability to attach a new volume using the same device name, unless you reboot the instance.
-  - We cannot reboot instance after each force-detach nor we can afford to &quot;loose&quot; a device name. AWS supports only 40 volumes per node and even that is quite low number already.
-- AWS.CreateVolume is not idempotent. There is no way to create a volume with ID provided by user. Such call would then fail when such volume already exists.
-- AWS.CreateVolume does not return errors when creating an encrypted volume using either non-existing or non-available KMS key (e.g. with wrong permission). It returns success instead and it even returns volumeID of some volume! This volume exists for a short while and it&#39;s deleted in couple of seconds.
 
-### Errors with slow kubelet
-
-Very rarely a node gets too busy and kubelet starves for CPU. It does not unmount a volume when it should and Kubernetes initiates detach of the volume.
+## EBS CSI Driver on Kubernetes
+
+### High Level Summary
+
+The EBS CSI Driver is used by Container Orchestrators to manage the lifecycle of AWS Elastic Block Storage Volumes. It is compliant to the [Container Storage Interface Specification](https://github.com/container-storage-interface/spec/blob/master/spec.md).
+
+On Kubernetes (K8s), the driver is split into two components: the ebs-csi-controller Deployment and ebs-csi-node DaemonSet. 
+
+The ebs-csi-controller watches K8s storage resources for changes, acts on those changes by making AWS EC2 API calls, and then updates those K8s resources once all associated EBS volumes reflect those changes. Inside the ebs-csi-controller pod are several CSI sidecar containers that interact with Kubernetes resources, as well as the ebs-plugin container which manages EBS Volumes via AWS API calls. The sidecars trigger these EBS volume operations by making CSI Remote Procedure Calls (RPCs) against the ebs-plugin's controller service. 
```

Comment:
nit: gRPC not RPC

Reply:
IMO gRPC is more of an implementation detail of what kind of RPC is made, but can change if folks think it is helpful.

Reply:
I suggest we stick with RPC except in the specific few sentences where it specifically matters that the protocal is gRPC.

As a guideline - if CSI were to switch to HTTP tomorrow, and all you would do for a sentence is replace "gRPC" with "HTTP RPC", then it should be "RPC" as the transport mechanism does not matter.

For the sections where it really does matter that it is gRPC, such as the sections explaining timeout handling, it should say gRPC.